src/ser.rs

```diff
@@ -1,9 +1,111 @@
+/// be applied. That means the results of writing two data blocks with one or two `BytesToBase32`
+/// converters will differ.
+struct BytesToBase32<'a, W: WriteBase32 + 'a> {
+	/// Target for writing the resulting `u5`s resulting from the written bytes
+	writer: &'a mut W,
+	/// Holds all unwritten bits left over from last round. The bits are stored beginning from
+	/// the most significant bit. E.g. if buffer_bits=3, then the byte with bits a, b and c will
+	/// look as follows: [a, b, c, 0, 0, 0, 0, 0]
+	buffer: u8,
+	/// Amount of bits left over from last round, stored in buffer.
+	buffer_bits: u8,
+}
+
+impl<'a, W: WriteBase32> BytesToBase32<'a, W> {
+	/// Create a new bytes-to-base32 converter with `writer` as  a sink for the resulting base32
+	/// data.
+	pub fn new(writer: &'a mut W) -> BytesToBase32<'a, W> {
+		BytesToBase32 {
+			writer,
+			buffer: 0,
+			buffer_bits: 0,
+		}
+	}
+
+	/// Add more bytes to the current conversion unit
+	pub fn append(&mut self, bytes: &[u8]) -> Result<(), W::Err> {
+		for b in bytes {
+			self.append_u8(*b)?;
+		}
+		Ok(())
+	}
+
+	pub fn append_u8(&mut self, byte: u8) -> Result<(), W::Err> {
+		// Write first u5 if we have to write two u5s this round. That only happens if the
+		// buffer holds too many bits, so we don't have to combine buffer bits with new bits
+		// from this rounds byte.
+		if self.buffer_bits >= 5 {
+			self.writer.write_u5(
+				u5::try_from_u8((self.buffer & 0b11111000) >> 3 ).expect("<32")
+			)?;
+			self.buffer = self.buffer << 5;
+			self.buffer_bits -= 5;
+		}
+
+		// Combine all bits from buffer with enough bits from this rounds byte so that they fill
+		// a u5. Save reamining bits from byte to buffer.
+		let from_buffer = self.buffer >> 3;
+		let from_byte = byte >> (3 + self.buffer_bits); // buffer_bits <= 4
+
+		self.writer.write_u5(u5::try_from_u8(from_buffer | from_byte).expect("<32"))?;
+		self.buffer = byte << (5 - self.buffer_bits);
+		self.buffer_bits = 3 + self.buffer_bits;
+
+		Ok(())
+	}
+
+	pub fn finalize(mut self) ->  Result<(), W::Err> {
+		self.inner_finalize()?;
+		std::mem::forget(self);
+		Ok(())
+	}
+
+	fn inner_finalize(&mut self) -> Result<(), W::Err>{
+		// There can be at most two u5s left in the buffer after processing all bytes, write them.
+		if self.buffer_bits >= 5 {
+			self.writer.write_u5(
+				u5::try_from_u8((self.buffer & 0b11111000) >> 3).expect("<32")
+			)?;
+			self.buffer = self.buffer << 5;
+			self.buffer_bits -= 5;
+		}
+
+		if self.buffer_bits != 0 {
+			self.writer.write_u5(u5::try_from_u8(self.buffer >> 3).expect("<32"))?;
+		}
+
+		// FIXME: finalize writer explicitly
+
+		Ok(())
+	}
+}
+
+impl<'a, W: WriteBase32> Drop for BytesToBase32<'a, W> {
+	fn drop(&mut self) {
+		self.inner_finalize()
+			.expect("Unhandled error when finalizing conversion on drop. User finalize to handle.")
+	}
+}
+
+/// Calculates the base32 encoded size of a byte slice
+fn bytes_size_to_base32_size(byte_size: usize) -> usize {
```

Comment:
Looks like the bech32 crate has an impl of this we can use?

Reply:
The code [exists](https://docs.rs/bech32/0.7.1/src/bech32/lib.rs.html#297-306) but isn't accessible as a public function. That should definitely be deduplicated when adding `BytesToBase32`.